docs(guides): Adds scaffolding (#2860)
webpack/webpack.js.org: src/content/guides/scaffolding.md

---
title: Scaffolding
contributors:
  - pranshuchittora
sort: 2
---

Setting up webpack for the first time is hard. Writing advanced configurations to optimize performance is even tougher. The `init` feature is designed to support people that want to create their own configuration or initializing other projects people create.

## Creating a scaffold

Before writing a `webpack-cli` scaffold, think about what you're trying to achieve. Do you want a "general" scaffold that could be used by any project or type of app? Do you want something focused, like a scaffold that writes both your `webpack.config.js` and your framework code? It's also useful to think about the user experience for your scaffold.

`webpack-cli` offers an experience that is interactive and you can prompt users for questions (like, "What is your entry point?") to help customize the output accordingly.

### Writing a scaffold

> There are many resources where you can learn how to write a scaffold, you can start from: [How do I compose a

Author: > Link to it

I'm sorry but I didn't it

webpack-scaffold?](https://github.com/evenstensberg/webpack-scaffold-demo)

`webpack-scaffold` is a utility suite for creating scaffolds. It contains functions that could be used to create a scaffold.

### Running a scaffold

A scaffold can be executed using `webpack-cli init`:

```bash
webpack-cli init <your-scaffold>
```

#### Running a scaffold locally

When the scaffold package is in your local file system you should pass its path to `init`:

```bash
webpack-cli init path/to/your/scaffold
```

Or you can create a global module and symlink to the local one:

- Using npm

      	```bash
      	cd path/to/my-scaffold
      	npm link
      	webpack-cli init my-scaffold
      	```

- Using yarn

      	```bash
      	cd path/to/my-scaffold
      	yarn link

>       	webpack-cli init my-scaffold

Author: oops!

#### Running a scaffold from npm

If the package is in npm, its name must begin with `webpack-scaffold` and can be used running:

```bash
webpack-cli init webpack-scaffold-yourpackage
```

## API

To create a `scaffold`, you must create a [`yeoman-generator`](http://yeoman.io/authoring/). Because of that, you can optionally extend your generator to include methods from the [Yeoman API](http://yeoman.io/learning/). It's worth noting that we support all the properties of a regular webpack configuration. In order for us to do this, there's a thing you need to remember:

> Objects are made using strings, while strings are made using double strings. This means that in order for you to create a string, you have to wrap it inside another string for us to validate it correctly.

### Required

- [opts.env.configuration](#optsenvconfiguration-required)
- [opts.env.configuration.myObj](#optsenvconfigurationmyObj-required)

> - [myObj.webpackOptions](#myObjwebpackOptions-required)

Reviewer: Using `config` is more intuitive instead of `myObj` IMO. We can have it changed everywhere.
pinging @evenstensberg for final call.

Author: I think it's might get confusing  as `opts.env.configuration.myObj` would become `opts.env.configuration.config`

Reviewer 2: In another PR Even suggested me not to change `myObj` to `myConfig` because it was confusing as sais Pranshu.
But I agree with Rishabh that it's not very intuitive, IMO would be nice to have a more meaningful name that gives the idea that it will become a webpack configuration file

Author: Yup, makes sense, what about `customConfig` or something like that.

Reviewer 2: I would also consider changing `opts.env.configuration` to `opts.env.configurations`

Reviewer 3: regarding configurations, yes. Regards to myObj is bc it sounds singular, not plural. You can nest the boostrap object however you want, needs to be emphasized if we're changing that.

That's all paper engineering. Submit an issue and let's go through that separately as in issue -> assignee -> pr -> release.

- [writing()](<#writing()-required>)

### Optional

- [myObj.merge](#myObjmerge-optional)
- [myObj.topScope](#myObjtopScope-optional)
- [myObj.configName](#myObjconfigName-optional)

> ### `opts.env.configuration`(required)

Reviewer 4: ```suggestion
### `opts.env.configuration`(required)

`object`
```

This is the entry point your configuration, initialize it inside the constructor of your generator in order for the CLI to work:

```js
class MyScaffold extends Generator {
  constructor(args, opts) {
    super(args, opts);
    opts.env.configuration = {};
  }
}
```

> ### `opts.env.configuration.myObj` (required)

Reviewer 4: ```suggestion
### `opts.env.configuration.myObj` (required)

`object`
```

This is your scaffold, you add here the options that the CLI will transform into a webpack configuration. You can have many different scaffolds named as you prefer, representing different configurations like `dev.config` or `prod.config`:

```js
class MyScaffold extends Generator {
  constructor(args, opts) {
    super(args, opts);
    opts.env.configuration = {
      dev: {},
      prod: {}
    };
  }
}
```

> ### `myObj.webpackOptions` (required)

Reviewer 4: ```suggestion
### `myObj.webpackOptions` (required)

`object`
```

> This object has the same format as a regular webpack configuration, so you declare here the properties that you want to scaffold, like `entry`, `output` and `context`. You can initialize this inside a yeoman method:

Reviewer 4: ```suggestion
This object has the same format as a regular webpack [configuration](/configuration/). Declare the properties that you want to scaffold here, e.g. `entry`, `output` and `context`. You can initialize this inside a yeoman method:
```

```js
this.options.env.configuration.dev.webpackOptions = {
  entry:'\'app.js\'',
  output:{}
};
```

> ### `myObj.merge` (optional)

Reviewer 4: ```suggestion
### `myObj.merge` (optional)

`string`
```

If you want to use [`webpack-merge`](https://github.com/survivejs/webpack-merge), you can set the `merge` property of `myObj` to the name of the configuration you want to merge it with:

```js
this.options.env.configuration.dev.merge = 'myConfig';
```

> ### `myObj.topScope`(optional)

Reviewer 4: ```suggestion
### `myObj.topScope`(optional)

`[string]`
```

> The `topScope` property is where you write all the code needed by your configuration, like module imports and functions/variables definitions:

Reviewer 4: ```suggestion
The `topScope` property is where you write all the code needed by your configuration, like module imports and functions/variables declarations:
```

```js
this.options.env.configuration.dev.topScope = [
  'const webpack = require("webpack");',
  'const path = require("path");'
];
```

> ### `myObj.configName`(optional)

Reviewer 4: ```suggestion
### `myObj.configName`(optional)

`string`
```

`configName` allows you to customize the name of your configuration file. For example you can name it `webpack.base.js` instead of the default `webpack.config.js`:

```js
this.options.env.configuration.dev.configName = 'base';
```

> ### `writing` (required)

Reviewer 4: ```suggestion
### `writing` (required)

`function`
```

## Creating a scaffold

Before writing a `webpack-cli` scaffold, think about what you're trying to achieve. Do you want a "general" scaffold that could be used by any project or type of app? Do you want something focused, like a scaffold that writes both your `webpack.config.js` and your framework code? It's also useful to think about the user experience for your scaffold.

`webpack-cli` offers an interactive experience to customize the output accordingly. For example asking questions like: "What is your entry point?".

### Writing a scaffold

There are many resources where you can learn how to write a scaffold, you can start from [Writing a Scaffold](/contribute/writing-a-scaffold/)

`webpack-scaffold` is a utility suite for creating scaffolds. It contains functions that could be used to create a scaffold.

### Running a scaffold

A scaffold can be executed using `webpack-cli init`:

```bash
webpack-cli init <your-scaffold>
```

#### Running a scaffold locally

When the scaffold package is in your local file system you should point `init` to its path:

```bash
webpack-cli init path/to/your/scaffold
```

Or you can create a global module and symlink to the local one:

- Using npm

>   `bash cd path/to/my-scaffold npm link webpack-cli init my-scaffold`

Reviewer 4: <img width="742" alt="Screen Shot 2019-03-25 at 11 25 30 AM" src="https://user-images.githubusercontent.com/10549495/54908501-b9edd580-4ef0-11e9-9518-768ae39dede3.png">
Does it look correct for you when the page is on website?

Reviewer 4: @pranshuchittora this doesnt look right

Author: Fixed
TIp:There must be an extra blank line surround the list items. 
```
./src/content/guides/scaffolding.md: 41: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "```bash"]
./src/content/guides/scaffolding.md: 42: MD031/blanks-around-fences Fenced code blocks should be surrounded by blank lines [Context: "```bash"]
./src/content/guides/scaffolding.md: 47: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "```bash"]
./src/content/guides/scaffolding.md: 48: MD031/blanks-around-fences Fenced code blocks should be surrounded by blank lines [Context: "```bash"]
```

- Using yarn
  `bash cd path/to/my-scaffold yarn link webpack-cli init my-scaffold`

#### Running a scaffold from npm

If the package is available from npm, its name must begin with `webpack-scaffold` and can be used by running:

```bash
webpack-cli init webpack-scaffold-yourpackage
```

## API

To create a `scaffold`, you must create a [`yeoman-generator`](http://yeoman.io/authoring/). Thanks to it, you can optionally extend your generator to include methods from the [Yeoman API](http://yeoman.io/learning/). It's worth noting that we support all the properties of a regular webpack configuration. In order for us to do this, there's a thing you need to remember:

W> Objects are made using strings, while strings are made using double strings. This means that in order for you to create a string, you have to wrap it inside another string for us to validate it correctly.

### Required

- [API](#api)

> 

Reviewer 4: Alignment is lost, please re-align

Author: markdown linter is throwing error, that's why it's like this.

Reviewer 4: Okay, i can resolve that later

---
title: Scaffolding
contributors:
  - evenstensberg
  - pranshuchittora
sort: 2
---

It can be hard to set up a complex webpack configuration for the first time. Writing advanced configurations to optimize performance is quite hard. The `init` feature is designed to support people that want to create their own configuration or initializing projects that other people create.

> ## Creating a scaffold

Reviewer: Merge this section under `Writing a scaffold` heading in this page at Line 17.

Before writing a `webpack-cli` scaffold, think about what you're trying to achieve. Do you want a "general" scaffold that could be used by any project or type of app? Do you want something focused, like a scaffold that writes both your `webpack.config.js` and your framework code? It's also useful to think about the user experience for your scaffold.

> `webpack-cli` offers interactive experience to customize the output accordingly. For example asking questions like: "What is your entry point?".

Reviewer 5: ```suggestion
`webpack-cli` offers an interactive experience to customize the output accordingly. For example asking questions like: "What is your entry point?".
```

### Writing a scaffold

> There are many resources where you can learn how to write a scaffold, you can start from: [Writing a Scaffold](/contribute/writing-a-scaffold/)

Reviewer 5: ```suggestion
There are many resources where you can learn how to write a scaffold, you can start from [Writing a Scaffold](/contribute/writing-a-scaffold/)
```

---
title: Scaffolding
contributors:
  - evenstensberg
  - pranshuchittora
sort: 2
---

It can be hard to set up a complex webpack configuration for the first time. Writing advanced configurations to optimize performance is quite hard. The `init` feature is designed to support people that want to create their own configuration or initializing projects that other people create.

## Creating a scaffold

Before writing a `webpack-cli` scaffold, think about what you're trying to achieve. Do you want a "general" scaffold that could be used by any project or type of app? Do you want something focused, like a scaffold that writes both your `webpack.config.js` and your framework code? It's also useful to think about the user experience for your scaffold.

`webpack-cli` offers interactive experience to customize the output accordingly. For example asking questions like: "What is your entry point?".

### Writing a scaffold

> There are many resources where you can learn how to write a scaffold, you can start from: [Writing a Scaffold](/contribute/writing-a-scaffold)

Reviewer 4: Need this for dev-server to open this locally
```suggestion
There are many resources where you can learn how to write a scaffold, you can start from: [Writing a Scaffold](/contribute/writing-a-scaffold/)
```

`webpack-scaffold` is a utility suite for creating scaffolds. It contains functions that could be used to create a scaffold.

### Running a scaffold

A scaffold can be executed using `webpack-cli init`:

```bash
webpack-cli init <your-scaffold>
```

#### Running a scaffold locally

When the scaffold package is in your local file system you should point `init` to its path:

```bash
webpack-cli init path/to/your/scaffold
```

Or you can create a global module and symlink to the local one:

- Using npm

>       	```bash

Reviewer 4: Please remove extra spacing before before code block lines for npm and yarn examples

`webpack-cli` offers an interactive experience to customize the output accordingly. For example asking questions like: "What is your entry point?".

### Writing a scaffold

There are many resources where you can learn how to write a scaffold, you can start from [Writing a Scaffold](/contribute/writing-a-scaffold/)

`webpack-scaffold` is a utility suite for creating scaffolds. It contains functions that could be used to create a scaffold.

### Running a scaffold

A scaffold can be executed using `webpack-cli init`:

```bash
webpack-cli init <your-scaffold>
```

#### Running a scaffold locally

When the scaffold package is in your local file system you should point `init` to its path:

```bash
webpack-cli init path/to/your/scaffold
```

Or you can create a global module and symlink to the local one:

- Using npm

```bash
cd path/to/my-scaffold

> npm link webpack-cli init my-scaffold

Reviewer 4: that wont run, go with
```suggestion
npm link
webpack-cli init my-scaffold
```

```

- Using yarn

```bash
cd path/to/my-scaffold

> yarn link webpack-cli init my-scaffold

Reviewer 4: same here:
```suggestion
yarn link
webpack-cli init my-scaffold
```
